src/target/rp.c

```diff
@@ -554,56 +611,216 @@ static void rp_spi_read(
+	target_mem_write32(t, RP_GPIO_QSPI_SD3_CTRL, 0);
+}
 
-	if (i < 16) {
-		// We have some data (hopefully a valid program) stored in the start
-		// of the flash memory. We can check if the start of this data is
-		// mirrored anywhere else in the flash as the flash region will repeat
-		// when we try to read out of bounds.
-		uint32_t mirrorsec[16];
-		while (size > FLASHSIZE_4K_SECTOR) {
-			target_mem_read(t, mirrorsec, RP_XIP_FLASH_BASE + size, sizeof(bootsec));
-			if (memcmp(bootsec, mirrorsec, sizeof(bootsec)) != 0)
-				return size << 1U;
-			size >>= 1U;
-		}
-	}
+// Set up the SSI controller for standard SPI mode,i.e. for every byte sent we get one back
+// This is only called by flash_exit_xip(), not by any of the other functions.
+// This makes it possible for the debugger or user code to edit SPI settings
+// e.g. baud rate, CPOL/CPHA.
+static void rp_flash_init_spi(target *t)
+{
+	// Disable SSI for further config
+	target_mem_write32(t, RP_SSI_ENABLE, 0);
+	// Clear sticky errors (clear-on-read)
+	target_mem_read32(t, RP_SSI_SR);
+	target_mem_read32(t, RP_SSI_ICR);
+	// Hopefully-conservative baud rate for boot and programming
+	target_mem_write32(t, RP_SSI_BAUD, 6);
+	target_mem_write32(t, RP_SSI_CTRL0,
+			RP_SSI_CTRL0_FRF_SERIAL |	// Standard 1-bit SPI serial frames
+			RP_SSI_CTRL0_DATA_BITS(8) |	// 8 clocks per data frame
+			RP_SSI_CTRL0_TMOD_BIDI		// TX and RX FIFOs are both used for every byte
+	);
+	// Slave selected when transfers in progress
+	target_mem_write32(t, RP_SSI_SER, 1);
+	// Re-enable
+	target_mem_write32(t, RP_SSI_ENABLE, 1);
+}
+
+// Also allow any unbounded loops to check whether the above abort condition
+// was asserted, and terminate early
+static int rp_flash_was_aborted(target *t) {
+    return target_mem_read32(t, RP_GPIO_QSPI_SD1_CTRL) & RP_GPIO_QSPI_SD1_CTRL_INOVER_BITS;
+}
 
-	// That approach didn't work. Most likely because there was no data found in
-	// at the start of the flash memory. If we have no valid program it's ok to
-	// interrupt the flash execution to check the JEDEC ID of the flash chip.
-	size = MAX_FLASH;
+// Put bytes from one buffer, and get bytes into another buffer.
+// These can be the same buffer.
+// If tx is NULL then send zeroes.
+// If rx is NULL then all read data will be dropped.
+//
+// If rx_skip is nonzero, this many bytes will first be consumed from the FIFO,
+// before reading a further count bytes into *rx.
+// E.g. if you have written a command+address just before calling this function.
+static void rp_flash_put_get(target *t, const uint8_t *tx, uint8_t *rx, size_t count, size_t rx_skip) {
+    // Make sure there is never more data in flight than the depth of the RX
+    // FIFO. Otherwise, when we are interrupted for long periods, hardware
+    // will overflow the RX FIFO.
+    const uint max_in_flight = 16 - 2; // account for data internal to SSI
+    size_t tx_count = count;
+    size_t rx_count = count;
+    while (tx_count || rx_skip || rx_count) {
```

Comment:
As much as this is a reasonably complex condition, is it possible to turn this into a for loop, allowing tx and rx to be marked as const pointers? (`const uint8_t *const tx, uint8_t *const rx`)

It's our experience that this kind of while loop that's really a for loop both hurts codegen and so also Flash usage while being harder to reason about.

Reply:
It might be possible, but I feel that there is a large amount of clarity gained by using a while loop, over using a for loop.
The number of times the loop should be run is not possible to be determined before hand as it depends on the levels read from `RP_SSI_TXFLR` and `RP_SSI_RXFLR`.
But I would be happy to accept suggestions for changes!

Reply:
So.. this does make the result 8 bytes larger in Flash (not entirely sure why).. but.. this is what we meant about finding this complex and hard to follow so could it be turned into a for loop:

```c
static void rp_flash_put_get(target *const t, const uint8_t *const tx, uint8_t *const rx,
	const size_t count, const size_t rx_skip)
{
	// Make sure there is never more data in flight than the depth of the RX
	// FIFO. Otherwise, when we are interrupted for long periods, hardware
	// will overflow the RX FIFO.
	static const size_t max_in_flight = 16 - 2; // account for data internal to SSI
	size_t rx_offset = 0;
	for (size_t tx_count = 0, rx_count = 0; rx_count < count + rx_skip || tx_count < count; ) {
		// NB order of reads, for pessimism rather than optimism
		const uint32_t tx_level = target_mem_read32(t, RP_SSI_TXFLR);
		const uint32_t rx_level = target_mem_read32(t, RP_SSI_RXFLR);
		bool idle = true; // Expect this to be folded into control flow, not register

		if (tx_count < count && tx_level + rx_level < max_in_flight) {
			target_mem_write32(t, RP_SSI_DR0, tx ? tx[tx_count] : 0U);
			++tx_count;
			idle = false;
		}

		if (rx_level) {
			const uint8_t data = target_mem_read32(t, RP_SSI_DR0) & 0xFFU;
			if (rx_count >= rx_skip && rx)
				rx[rx_offset++] = data;
			++rx_count;
			idle = false;
		}

		if (idle && rp_flash_was_aborted(t))
			break;
	}
	rp_spi_chip_select(t, RP_GPIO_QSPI_CS_DRIVE_HIGH);
}
```

It's still a bit messy to follow, but we find both because it's doing less shenanigans with count-down parameters, and because of the reduced indentation levels, the trade off is worth it. We'll let you have final say but this would be our suggestion.

Reply:
I think for clarity it should be left as a while loop as the step size for the loop is not constant and therefore does not align with the for loop idiom. But I have updated the function to make the `tx` and `rx` const pointers, and performed a little bit more clean-up, so I'm happy now!